### PR TITLE
Go: Implement `stdin` models

### PR DESCRIPTION
**Pull request by egregius313**
Adds source models for the `stdin` threat model.

These models are implemented in QL instead of MaD due to limitations in MaD:
- MaD does not currently handle variable arguments as well as `getAnImplicitVarargsArgument`
- MaD has a way to access fields on structs, but not variables in a package. (This is generally not that important, but it is necessary for modeling `os.Stdin`).

### CODE REVIEW

File: go/ql/lib/semmle/go/frameworks/stdlib/Os.qll [hunk 1]
```diff
@@ -43,4 +43,14 @@ module Os {
       input = inp and output = outp
     }
   }
+
+  private class Stdin extends SourceNode {
+    Stdin() {
+      exists(Variable osStdin | osStdin.hasQualifiedName("os", "Stdin") |
+        this.asExpr() = osStdin.getARead().asExpr()
```
Comment (owen-mc) [inline go/ql/lib/semmle/go/frameworks/stdlib/Os.qll:50]:
```suggestion
        this = osStdin.getARead()
```
There are two kinds of read instructions which aren't expressions. Neither seems particularly relevant here, but it seems easier to not restrict to expressions.

File: go/ql/lib/change-notes/2024-08-29-add-stdin-threat-models.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+* Local source models with the `stdin` source kind have been added for `os.Stdin`, `fmt.Scan`, `fmt.Scanf`, `fmt.Scanln`. You can optionally include threat models as appropriate when using the CodeQL CLI and in GitHub code scanning. For more information, see [Analyzing your code with CodeQL queries](https://docs.github.com/code-security/codeql-cli/getting-started-with-the-codeql-cli/analyzing-your-code-with-codeql-queries#including-model-packs-to-add-potential-sources-of-tainted-data>) and [Customizing your advanced setup for code scanning](https://docs.github.com/code-security/code-scanning/creating-an-advanced-setup-for-code-scanning/customizing-your-advanced-setup-for-code-scanning#extending-codeql-coverage-with-threat-models).
```
Comment (owen-mc) [inline go/ql/lib/change-notes/2024-08-29-add-stdin-threat-models.md:4]:
```suggestion
* Local source models with the `stdin` source kind have been added for the variable `os.Stdin` and the functions `fmt.Scan`, `fmt.Scanf` and `fmt.Scanln`. You can optionally include threat models as appropriate when using the CodeQL CLI and in GitHub code scanning. For more information, see [Analyzing your code with CodeQL queries](https://docs.github.com/code-security/codeql-cli/getting-started-with-the-codeql-cli/analyzing-your-code-with-codeql-queries#including-model-packs-to-add-potential-sources-of-tainted-data>) and [Customizing your advanced setup for code scanning](https://docs.github.com/code-security/code-scanning/creating-an-advanced-setup-for-code-scanning/customizing-your-advanced-setup-for-code-scanning#extending-codeql-coverage-with-threat-models).
```

**Review by owen-mc — DISMISSED** [review]
Looks good. Only trivial comments.

**Comment by owen-mc** [comment]
Looks like my suggestion needs reformatting.

File: go/ql/test/library-tests/semmle/go/dataflow/flowsources/local/stdin/test.ql [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import go
+import semmle.go.dataflow.ExternalFlow
+import ModelValidation
+import experimental.frameworks.CleverGo
+import TestUtilities.InlineFlowTest
+
+module Config implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) { source instanceof ThreatModelFlowSource }
```
Comment (egregius313) [inline go/ql/test/library-tests/semmle/go/dataflow/flowsources/local/stdin/test.ql:8]:
```suggestion
  predicate isSource(DataFlow::Node source) { source instanceof ActiveThreatModelSource }
```

File: go/ql/test/library-tests/semmle/go/dataflow/flowsources/local/stdin/source.ql [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import go
+import ModelValidation
+import TestUtilities.InlineExpectationsTest
+
+module SourceTest implements TestSig {
+  string getARelevantTag() { result = "source" }
+
+  predicate hasActualResult(Location location, string element, string tag, string value) {
+    exists(ThreatModelFlowSource s |
```
Comment (egregius313) [inline go/ql/test/library-tests/semmle/go/dataflow/flowsources/local/stdin/source.ql:9]:
```suggestion
    exists(ActiveThreatModelSource s |
```

**Pull request merged**